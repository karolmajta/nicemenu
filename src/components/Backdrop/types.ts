import { ReactChild } from "react";

export interface BackdropProps {
  children: ReactChild;
  className?: string;
  classNames?: Array<string>;
  onDismiss: () => void;
}
