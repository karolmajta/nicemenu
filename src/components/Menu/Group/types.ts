import { IconName } from "../Item/Icon/types";
import React from "react";
import { AllowedShortcutKey, ItemProps } from "../Item/types";

export interface GroupItemProps {
  icon: IconName;
  label: string;
  shortcut?: Array<Array<AllowedShortcutKey>>;
  goto?: string;
  onSelect?: () => void;
  listIndex: number;
}

export interface GroupProps {
  label: string;
  items: Array<GroupItemProps>;
  activeListIndex: number | null;
  className?: string;
  classNames?: Array<string>;
  components?: {
    Item?: React.ComponentType<ItemProps>;
  };
  onItemMouseMove?: (index: number) => void;
  onGoto: (gotoPath: string) => void;
}
