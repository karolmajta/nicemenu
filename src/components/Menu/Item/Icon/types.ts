export type IconName =
  | "bolt"
  | "dot"
  | "inbox"
  | "loader"
  | "pencil"
  | "reschedule";

export interface IconProps {
  name: IconName;
  className?: string;
  classNames?: Array<string>;
}
