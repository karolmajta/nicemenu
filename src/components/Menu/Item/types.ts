import React from "react";
import { IconName, IconProps } from "./Icon/types";

export type AllowedShortcutKey =
  | "a"
  | "b"
  | "c"
  | "d"
  | "e"
  | "f"
  | "g"
  | "h"
  | "i"
  | "j"
  | "k"
  | "l"
  | "m"
  | "n"
  | "o"
  | "p"
  | "q"
  | "r"
  | "s"
  | "t"
  | "u"
  | "v"
  | "w"
  | "x"
  | "y"
  | "z"
  | "Shift";

export interface ItemProps {
  label: string;
  icon?: IconName;
  shortcut?: Array<Array<AllowedShortcutKey>>;
  listIndex: number;
  highlighted: boolean;
  goto?: string;
  className?: string;
  classNames?: Array<string>;
  components?: {
    Icon?: React.ComponentType<IconProps>;
  };
  onMouseMove?: (listIndex: number) => void;
  onSelect?: () => void;
  onGoto: (menuKey: string) => void;
}
