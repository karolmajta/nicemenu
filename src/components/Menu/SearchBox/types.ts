export interface SearchBoxProps {
  placeholder: string;
  searchTerm: string;
  focused: boolean;
  className?: string;
  classNames?: Array<string>;
  onArrowUp: () => void;
  onArrowDown: () => void;
  onChange: (searchTerm: string) => void;
  onFocus: () => void;
}
