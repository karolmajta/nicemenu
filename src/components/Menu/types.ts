import { IconName } from "./Item/Icon/types";
import React from "react";
import { ItemProps, AllowedShortcutKey } from "./Item/types";
import { GroupProps } from "./Group/types";
import { SearchBoxProps } from "./SearchBox/types";

export interface MenuItem {
  icon: IconName;
  label: string;
  shortcut?: Array<Array<AllowedShortcutKey>>;
  goto?: string;
  onSelect?: () => void;
}

export interface MenuGroup {
  label: string;
  items: Array<MenuItem>;
}

export interface MenuProps {
  title: string;
  items: Array<MenuItem>;
  groups: Array<MenuGroup>;
  submenus: {
    [submenuName: string]: MenuProps;
  };
  className?: string;
  classNames?: Array<string>;
  components?: {
    Item?: React.ComponentType<ItemProps>;
    Group?: React.ComponentType<GroupProps>;
    SearchBox?: React.ComponentType<SearchBoxProps>;
  };
}
