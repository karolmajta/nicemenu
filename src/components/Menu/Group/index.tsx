import React from "react";
import cx from "classnames";
import styled from "styled-components";
import Item, { ItemUnstyled } from "../Item";
import { GroupProps } from "./types";

export function GroupUnstyled({
  label,
  items,
  activeListIndex,
  className,
  classNames,
  components,
  onItemMouseMove,
  onGoto,
}: GroupProps) {
  const ItemComponent = components?.Item || ItemUnstyled;
  return (
    <div
      data-testid="Laserfocus-Menu-Group"
      className={cx("Laserfocus-Menu-Group", className, ...(classNames || []))}
    >
      <div className="group-label">
        <div className="group-label-text">{label}</div>
      </div>
      {items.map(({ icon, label, shortcut, listIndex, goto, onSelect }, i) => (
        <ItemComponent
          key={i}
          icon={icon}
          label={label}
          shortcut={shortcut}
          listIndex={listIndex}
          highlighted={listIndex === activeListIndex}
          onMouseMove={() => onItemMouseMove && onItemMouseMove(listIndex)}
          goto={goto}
          onGoto={onGoto}
          onSelect={() => onSelect && onSelect()}
        />
      ))}
    </div>
  );
}

const Group = styled(GroupUnstyled).attrs((props) => ({
  ...props,
  components: {
    Item: Item,
  },
}))`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;

  width: 100%;

  .group-label {
    display: flex;
    flex-direction: row;
    justify-content: flex-start;
    align-items: center;

    width: 100%;
    height: 36px;

    font-size: 12px;
    color: rgba(228, 228, 238, 0.5);

    background: rgba(228, 228, 238, 0.05);

    .group-label-text {
      display: block;

      margin-left: 20px;
    }
  }
`;

export default Group;
