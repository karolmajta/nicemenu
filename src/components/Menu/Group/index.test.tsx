import React from "react";
import { render, cleanup } from "@testing-library/react";
import Group from ".";
import { GroupProps } from "./types";

const renderGroup = (props: Partial<GroupProps>) =>
  render(
    <Group
      label={props.label || "Example label"}
      items={
        props.items || [{ label: "Eample item 1" }, { label: "Eample item 2" }]
      }
      activeListIndex={props.activeListIndex || 0}
      onGoto={props.onGoto || jest.fn()}
    />
  );

afterEach(() => {
  cleanup();
});

test("renders correctly", () => {
  const { getByTestId } = renderGroup({});
  const groupDiv = getByTestId("Laserfocus-Menu-Group");
  expect(groupDiv).toMatchSnapshot();
});

test("renders specified items", () => {
  const { getByTestId } = renderGroup({});
  const groupDiv = getByTestId("Laserfocus-Menu-Group");
  expect(
    groupDiv.querySelectorAll('[data-testid="Laserfocus-Menu-Item"]').length
  ).toEqual(2);
});
