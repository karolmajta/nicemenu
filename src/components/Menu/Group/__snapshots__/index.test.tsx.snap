// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="Laserfocus-Menu-Group sc-dlnjPT fMMgoZ"
  data-testid="Laserfocus-Menu-Group"
>
  <div
    class="group-label"
  >
    <div
      class="group-label-text"
    >
      Example label
    </div>
  </div>
  <div
    class="Laserfocus-Menu-Item sc-gtssRu iLlUnN"
    data-testid="Laserfocus-Menu-Item"
  >
    <div
      class="icon-placeholder"
    />
    <div
      class="item-label"
    >
      Eample item 1
    </div>
  </div>
  <div
    class="Laserfocus-Menu-Item sc-gtssRu iLlUnN"
    data-testid="Laserfocus-Menu-Item"
  >
    <div
      class="icon-placeholder"
    />
    <div
      class="item-label"
    >
      Eample item 2
    </div>
  </div>
</div>
`;
